authentik/core/models.py

```diff
@@ -586,6 +597,76 @@ def ui_user_settings(self) -> UserSettingSerializer | None:
+        """Get base properties for a user to build final properties upon."""
+        raise NotImplementedError
+
+    def get_base_group_properties(self, **kwargs) -> dict[str, Any | dict[str, Any]]:
+        """Get base properties for a group to build final properties upon."""
+        raise NotImplementedError
+
+    def get_base_properties(
+        self, object_type: type[User | Group], **kwargs
+    ) -> dict[str, Any | dict[str, Any]]:
+        """Get base properties for a user or a group to build final properties upon."""
+        if object_type == User:
+            properties = self.get_base_user_properties(**kwargs)
+            properties.setdefault("path", self.get_user_path())
+            return properties
+        if object_type == Group:
+            return self.get_base_group_properties(**kwargs)
+        return {}
+
+    def build_object_properties(
+        self,
+        object_type: type[User | Group],
+        user: User | None = None,
+        request: HttpRequest | None = None,
+        **kwargs,
+    ) -> dict[str, Any | dict[str, Any]]:
+        """Build a user or group properties from the source configured property mappings."""
+        from authentik.events.models import Event, EventAction
+
+        properties = self.get_base_properties(object_type, **kwargs)
+        if "attributes" not in properties:
+            properties["attributes"] = {}
+        mappings = []
+        if object_type == User:
+            mappings = self.user_property_mappings.all().select_subclasses()
+        elif object_type == Group:
+            mappings = self.group_property_mappings.all().select_subclasses()
+        print(mappings)
+        for mapping in mappings:
+            if not isinstance(mapping, self.property_mapping_type):
+                continue
+            try:
+                value = mapping.evaluate(
+                    user=user,
+                    request=request,
+                    source=self,
+                    properties=properties,
+                    **kwargs,
+                )
+                if not value or not isinstance(value, dict):
+                    LOGGER.debug(
+                        "Mapping evaluated to None or is not a dict. Skipping",
+                        source=self,
+                        mapping=mapping,
+                    )
+                    continue
+            except PropertyMappingExpressionException as exc:
+                Event.new(
+                    EventAction.CONFIGURATION_ERROR,
+                    message=f"Failed to evaluate property mapping: '{mapping.name}'",
+                    source=self,
+                    mapping=mapping,
+                ).save()
+                LOGGER.warning("Mapping failed to evaluate", exc=exc, source=self, mapping=mapping)
+                continue
+            MERGE_LIST_UNIQUE.merge(properties, value)
+
+        return remove_none(flatten_rec(properties))
```

Comment:
flatten_rec probably is a bad idea here. It should be handled in the LDAP source directly.

remove_none is a duplicate of authentik.policies.utils.delete_none, except that it is recursive.